### PR TITLE
Copy ingest

### PR DESCRIPTION
**Pull request by j23414**
## Description of proposed changes

Splitting out [dengue ingest changes PR#6](https://github.com/nextstrain/dengue/pull/6) into smaller, more focused pull requests. Especially since there have been significant improvements to linking reusable pathogen ingest scripts.

The primary scope of this PR includes:

1. Copying the Monkeypox ingest directory.
2. Using git subtree to copy and link the reusable scripts in an `ingest/vendored` subdirectory.
3. Temporarily removing Nextclade-related rules, pending the compilation of a Nextclade dengue dataset and potential v3 changes.
4. Pulling and processing one "sequences.fasta" and "metadata.tsv" file pair.

To ease development and review, tasks that are not part of this PR but will be submitted as future PRs include:

1. Splitting the fetch process into dengue serotypes (denv1-4).
2. Adding dengue-specific annotations and data fixes.
3. Integrating Nextclade-related rules and datasets.

## Related issue(s)

* https://github.com/nextstrain/dengue/pull/6
* https://github.com/nextstrain/ingest/issues/9

## Checklist

- [ ] Checks pass
- [ ] Manual check passes

```
git clone https://github.com/nextstrain/dengue.git
cd dengue
git checkout cp_ingest
cd ingest
nextstrain build . data/metadata.tsv data/sequences.fasta
```

### CODE REVIEW

**Review by jameshadfield — CHANGES_REQUESTED** [review]
I had to install `ncbi-datasets-cli` to run this (missing `datasets` command) - it's listed in our [conda recipe](https://github.com/nextstrain/conda-base/blob/main/src/recipe.yaml#L75C7-L75C24) but it's not listed in the "ambient" section of [our install docs](https://docs.nextstrain.org/en/latest/install.html). I'd suggest it's added to those docs, or detailed in the README here; even better would be to catch the error and explain what piece of software is missing to the user.

Using ncbi-datasets-cli version 15.23.0 I encountered the following error: `Error: Internal error (invalid zip archive). Please try again` as part of rule `fetch_ncbi_dataset_package`. Perhaps I'm missing some env variables? Perhaps it's transient (although it happened multiple times)?

(I don't have a view on whether copy-and-pasting the mpx ingest is the right approach; I'll defer to @joverlee521 here.)

**Comment by j23414** [comment]
> I had to install `ncbi-datasets-cli`

Ah I see what happened. It appears we switched from NCBI Virus to NCBI Datasets on [Sept 11, 2023](82ace30abd07bec140c7e14aee29a46c7bfb28a8) and the ingest README here and in [Monkeypox]( https://github.com/nextstrain/monkeypox/tree/master/ingest#input-data) need to be updated. I'll ensure that the ingest README for Monkeypox is updated first, followed by a rebase of this PR. Additionally, I'll submit a PR to update the ambient documentation.

> Error: Internal error (invalid zip archive)

I've encountered this error during testing in the Monkeypox repository as well. It seems to be a [known intermittent issue with ncbi datasets](https://github.com/ncbi/datasets/issues/89) . To mitigate this, we have a  [`retries: 5` directive in the Snakemake rule](https://github.com/nextstrain/monkeypox/blob/afb5513d9dea10e1993973a154844bf30cdbd335/ingest/workflow/snakemake_rules/fetch_sequences.smk#L20C52-L20C52). 

(For documentation:
The copy-and-pasting mpx ingest approach serves several benefits, including maintaining consistency across pathogen repositories, facilitating focused review on new changes, and providing a method to highlight changes that need to be propagated to other repositories.)

File: ingest/Snakefile
Comment (joverlee521) [inline ingest/Snakefile:17]: _nitpick (on my past self)_  

I'm not sure why I set up the monkeypox ingest workflow to output the final results to the `data` directory, maybe I carried it over from the ncov-ingest pipeline. 

I think it'd make more sense for final outputs to be in the `results` directory instead of being mixed with the intermediate data files.

File: ingest/workflow/snakemake_rules/trigger_rebuild.smk
Comment (joverlee521) [inline ingest/workflow/snakemake_rules/trigger_rebuild.smk:12]: Since writing this workflow, we've decided as team to [avoid the `message` attribute in Snakemake rules](https://docs.nextstrain.org/en/latest/reference/snakemake-style-guide.html#avoid-the-message-rule-attribute).

**Review by joverlee521 — CHANGES_REQUESTED** [review]
Thanks for breaking down dengue/ingest changes into smaller focused PRs @j23414, this is much easier to review!

> I don't have a view on whether copy-and-pasting the mpx ingest is the right approach; I'll defer to @joverlee521 here.

I think it's _fine_ to copy monkeypox/ingest here since it's based on previous work that Jennifer had done. I've made comments on some changes that would be small improvements upon the monkeypox workflow.

This does bring the question to mind of whether it's easier for the team to copy/paste an existing ingest workflow rather than start from scratch with the [pathogen-repo-template](https://github.com/nextstrain/pathogen-repo-template).

File: ingest/bin/reverse_reversed_sequences.py
Comment (j23414) [reply ingest/bin/reverse_reversed_sequences.py:1]: Fixed with https://github.com/nextstrain/dengue/pull/13/commits/0b2429b214167f2b331268bca09b68c3d7fee318

File: ingest/config/config.yaml
Comment (j23414) [reply ingest/config/config.yaml]: Fixed with https://github.com/nextstrain/dengue/pull/13/commits/0b2429b214167f2b331268bca09b68c3d7fee318

File: ingest/workflow/snakemake_rules/trigger_rebuild.smk
Comment (j23414) [reply ingest/workflow/snakemake_rules/trigger_rebuild.smk]: Fixed with https://github.com/nextstrain/dengue/pull/13/commits/a4d9ca74a43ac8bb3250f623d8272b76de9d281b and a separate PR to mkpx: https://github.com/nextstrain/monkeypox/pull/213

Later on, I expect this commit to disappear when I perform a rebase on mkpx ingest.

File: ingest/source-data/annotations.tsv
Comment (j23414) [reply ingest/source-data/annotations.tsv:1]: Fixed with 224c54532110c8ca76d211c8ea7ad3b97d140596

File: ingest/source-data/geolocation-rules.tsv
Comment (j23414) [reply ingest/source-data/geolocation-rules.tsv:1]: Fixed with 224c54532110c8ca76d211c8ea7ad3b97d140596

File: ingest/config/optional.yaml
Comment (j23414) [reply ingest/config/optional.yaml]: Fixed with 493507eda6922874e47763afd2471381baf1f610

File: ingest/config/optional.yaml
Comment (j23414) [reply ingest/config/optional.yaml]: Gah, can't believe I missed that one. Thanks! Fixed with f1e015ce005c2423d63b15ad1eb335a3f252c01d

File: ingest/README.md
Comment (j23414) [reply ingest/README.md]: Fixed by rebasing to mkpx ingest on 2023-10-25

File: ingest/config/optional.yaml
Comment (corneliusroemer) [reply ingest/config/optional.yaml]: This sounds like a good thing to add to our Snakemake styleguide: benefit of not nesting configs

File: ingest/source-data/ncbi-dataset-field-map.tsv
Comment (corneliusroemer) [inline ingest/source-data/ncbi-dataset-field-map.tsv:8]:
```suggestion
Release date	released
```
Pet peeve of mine to eradicate our confounding of release and submission ;)

File: ingest/source-data/ncbi-dataset-field-map.tsv
Comment (corneliusroemer) [inline ingest/source-data/ncbi-dataset-field-map.tsv:13]:
```suggestion
BioProjects	bioproject_accessions
```

File: ingest/bin/fasta-to-ndjson
Comment (corneliusroemer) [inline ingest/bin/fasta-to-ndjson:1]: Should this be vendored?

File: ingest/bin/ndjson-to-tsv-and-fasta
Comment (corneliusroemer) [inline ingest/bin/ndjson-to-tsv-and-fasta:1]: Also a candidate for vendoring?

File: ingest/config/config.yaml
Comment (corneliusroemer) [inline ingest/config/config.yaml:16]:
```suggestion
  date_fields: ['date', 'date_released']
```

File: ingest/config/config.yaml
Comment (corneliusroemer) [inline ingest/config/config.yaml:62]:
```suggestion
    'date_released',
```

File: ingest/source-data/ncbi-dataset-field-map.tsv
Comment (corneliusroemer) [inline ingest/source-data/ncbi-dataset-field-map.tsv:1]: Should we move this into the config file?

File: ingest/Snakefile
Comment (corneliusroemer) [inline ingest/Snakefile:3]: Minimum might actually be higher now that we use a default profile - something like 7.30

Below one might possibly need to pass in extra cli flags

File: ingest/workflow/snakemake_rules/fetch_sequences.smk
Comment (corneliusroemer) [inline ingest/workflow/snakemake_rules/fetch_sequences.smk:25]: The taxon id could be factored out into config

File: ingest/workflow/snakemake_rules/fetch_sequences.smk
Comment (corneliusroemer) [inline ingest/workflow/snakemake_rules/fetch_sequences.smk:72]: There's a lot of duplication of these mappings here - it might be nice to deduplicate

File: ingest/bin/fasta-to-ndjson
Comment (victorlin) [reply ingest/bin/fasta-to-ndjson:1]: It was considered for centralization but in the end [slated for replacement/removal](https://github.com/nextstrain/ingest/issues/28).

File: ingest/bin/ndjson-to-tsv-and-fasta
Comment (victorlin) [reply ingest/bin/ndjson-to-tsv-and-fasta:1]: Same as https://github.com/nextstrain/dengue/pull/13#discussion_r1373793587

File: ingest/source-data/ncbi-dataset-field-map.tsv
Comment (j23414) [reply ingest/source-data/ncbi-dataset-field-map.tsv:1]: I somewhat appreciate the renaming because I'm not a fan of the spaces and capitalization in the  NCBI Dataset field names (e.g. "Isolate Lineage" is being renamed to "strain" and "Geographic Region" to "region"). 

Do you prefer "Isolate Lineage" or would find "isolate_lineage" also acceptable?

Comment (joverlee521) [reply ingest/source-data/ncbi-dataset-field-map.tsv:1]: Hmm, I'd rather keep the fields exactly as NCBI has them. This way we can point users to the [NCBI Dataset docs for a description of each field](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/reference-docs/data-reports/virus/#ncbi-datasets-v2-reports-VirusAssembly). I think it would help us avoid things like the [submitted/released field mix up](https://github.com/nextstrain/dengue/pull/13#discussion_r1373356935) in the NDJSON.

File: ingest/config/optional.yaml
Comment (victorlin) [reply ingest/config/optional.yaml]: It's not very straightforward, but would something like [this](https://stackoverflow.com/a/66470308) work?

```
--config s3='{dst: "s3://nextstrain-data/files/workflows/monkeypox"}'
```

Comment (joverlee521) [reply ingest/config/optional.yaml]: OH! I just tested that the `--config` option works with 
```
--config upload='{"s3": {"dst": "NEW URL"}}'
```

Snakemake updated the config overwrite behavior in https://github.com/snakemake/snakemake/pull/1126, which was released as part of v6.11.0. So I guess nested configs is no longer that big of an issue 🎉

File: ingest/source-data/ncbi-dataset-field-map.tsv
Comment (j23414) [reply ingest/source-data/ncbi-dataset-field-map.tsv:1]: I really like the idea of pointing users to the NCBI Dataset field documentation! However, it is worth noting that the field `Geographic Region` is documented as `geographicRegion` and `Region`. I guess that would be NCBI's responsibility to keep their documentation up-to-date. 

Spaces in column names can require special handling (e.g. properly quoting column names) which I think our code has been made to handle. However, such spaces may post challenges in other languages, scripts, or projects. Replacing spaces with underscores in the NCBI Dataset fields, would still enable us to point to NCBI's documentation.

Nevertheless, I'm open to using dengue as a test pipeline to assess how our code handles columns with spaces. If needed, we can later reintroduce modifications to column titles.

File: ingest/workflow/snakemake_rules/fetch_sequences.smk
Comment (j23414) [reply ingest/workflow/snakemake_rules/fetch_sequences.smk:25]: Agree, submitted a PR to monkeypox https://github.com/nextstrain/monkeypox/pull/217
Will rebase those changes here after it's approved and merged.

File: ingest/source-data/ncbi-dataset-field-map.tsv
Comment (joverlee521) [reply ingest/source-data/ncbi-dataset-field-map.tsv:1]: > However, it is worth noting that the field Geographic Region is documented as geographicRegion and Region. 

Oh right, I forgot that the NCBI reference docs have a nonintuitive set up. They have a top level "Geographic" column name that gets prepended to the [VirusAssembly.CollectionLocation](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/reference-docs/data-reports/virus/#ncbi-datasets-v2-reports-VirusAssembly-CollectionLocation) fields. The `dataformat tsv virus-genome` [docs for the available fields](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/reference-docs/command-line/dataformat/tsv/dataformat_tsv_virus-genome/#fields) is a better place to see the full field names. 

So I guess there's no that much benefit from keeping the space matching for the column names...Welp, maybe we should transform the names back to the [mnemonics](https://github.com/nextstrain/dengue/blob/9961ec1b839d39fa80385ec935180c0350eb0af7/ingest/workflow/snakemake_rules/fetch_sequences.smk#L55-L72) so that we only have to maintain one set of column mappings.

File: ingest/config/optional.yaml
Comment (joverlee521) [reply ingest/config/optional.yaml]: See example of the config override in the [monkeypox fetch-and-ingest-branch workflow](https://github.com/nextstrain/monkeypox/pull/218/commits/73bbe5c2ba0bbbb1018210c8ee42fd22efc9350c)

File: ingest/source-data/ncbi-dataset-field-map.tsv
Comment (j23414) [reply ingest/source-data/ncbi-dataset-field-map.tsv:1]: Agree, working on it in https://github.com/nextstrain/monkeypox/pull/219

Comment (j23414) [reply ingest/source-data/ncbi-dataset-field-map.tsv:1]: Merged and rebased here

File: ingest/Snakefile
Comment (joverlee521) [reply ingest/Snakefile:17]: Updated upstream in [mpox repo](https://github.com/nextstrain/mpox/pull/222)

File: ingest/source-data/ncbi-dataset-field-map.tsv
Comment (j23414) [reply ingest/source-data/ncbi-dataset-field-map.tsv:1]: I made an attempt with https://github.com/nextstrain/dengue/pull/13/commits/c9caaab7af5d9c3d68ad91dd447e5d7037094a6d
I might be a bit uncertain on desired names for the final metadata, so any feedback is welcome.

Comment (joverlee521) [reply ingest/source-data/ncbi-dataset-field-map.tsv:1]: > I might be a bit uncertain on desired names for the final metadata, so any feedback is welcome.

I've been looking around for any documentation for standard Nextstrain metadata fields and found a page in the [ncov workflow docs](https://docs.nextstrain.org/projects/ncov/en/latest/reference/metadata-fields.html). I think we can use these standard metadata fields across all pathogens.

File: ingest/source-data/ncbi-dataset-field-map.tsv
Comment (j23414) [reply ingest/source-data/ncbi-dataset-field-map.tsv]: Ended up dropping the `BioProjects` column entirely since it is not used in the build (https://github.com/nextstrain/dengue/commit/116f363aa8ddba449720bc6feba2f7212390f781). I'll keep this comment in mind if we add it back in.

File: ingest/source-data/ncbi-dataset-field-map.tsv
Comment (j23414) [reply ingest/source-data/ncbi-dataset-field-map.tsv]: Agree, edited as part of commit: https://github.com/nextstrain/dengue/commit/116f363aa8ddba449720bc6feba2f7212390f781

File: ingest/source-data/ncbi-dataset-field-map.tsv
Comment (j23414) [reply ingest/source-data/ncbi-dataset-field-map.tsv:1]: [Looks like auspice requires `accession`](https://github.com/nextstrain/auspice/blob/477d27bc162a4f54f89e439a905b0f443d907a72/src/components/tree/infoPanels/click.js#L95C1-L98C7) instead of `genbank_accession` to render the link. Unless there's a workaround that I'm not seeing.

Screenshot 2023-11-13 at 4 30 52 PM

Comment (joverlee521) [reply ingest/source-data/ncbi-dataset-field-map.tsv:1]: [`gisaid_epi_isl` and `genbank_accession` are special cases in Auspice](https://github.com/nextstrain/auspice/blob/477d27bc162a4f54f89e439a905b0f443d907a72/src/components/tree/infoPanels/click.js#L56-L59)

Comment (j23414) [reply ingest/source-data/ncbi-dataset-field-map.tsv:1]: Just documenting that this PR is on hold until we resolve following questions: 
* [How do we want to standardize output metadata fields for Nextstrain ingest?](https://github.com/nextstrain/pathogen-repo-template/issues/20)
* [How do we want to configure that standardization in our pipelines?](https://github.com/nextstrain/pathogen-repo-template/pull/19)

**Review by joverlee521 — APPROVED** [review]
As I stated on [Slack](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1701283005269069?thread_ts=1701279901.637689&cid=C01LCTT7JNN), if the workflow in this PR is producing the public files on data.nextstrain.org, then I think it's fine to merge into the main branch now. 

We can tackle [unresolved questions](https://github.com/nextstrain/dengue/pull/13#discussion_r1408278053) in subsequent PRs.

**Pull request merged**
